src/back/spirv/parser.rs

```diff
@@ -0,0 +1,1545 @@
+        handle: &crate::Handle<crate::Type>,
+    ) -> Word {
+        match self.lookup_type.lookup_id(&handle) {
+            Some(word) => word,
+            None => {
+                let (instruction, id) = self.instruction_type_declaration(arena, *handle);
+
+                for words in instruction.to_words() {
+                    self.module.logical_layout.type_declarations.push(words);
+                }
+                id
+            }
+        }
+    }
+
+    fn get_constant_id(
+        &mut self,
+        handle: &crate::Handle<crate::Constant>,
+        ir_module: &crate::Module,
+    ) -> Word {
+        match self.lookup_constant.lookup_id(&handle) {
+            Some(word) => word,
+            None => {
+                let (instruction, id) = self.instruction_constant_type(*handle, ir_module);
+
+                for words in instruction.to_words() {
+                    self.module.logical_layout.constants.push(words);
+                }
+                id
+            }
+        }
+    }
+
+    fn get_global_variable_id(
+        &mut self,
+        arena: &crate::Arena<crate::Type>,
+        global_arena: &crate::Arena<crate::GlobalVariable>,
+        handle: &crate::Handle<crate::GlobalVariable>,
+    ) -> Word {
+        match self.lookup_global_variable.lookup_id(&handle) {
+            Some(word) => word,
+            None => {
+                let global_variable = &global_arena[*handle];
+                let (instruction, id) =
+                    self.instruction_global_variable(arena, global_variable, handle);
+
+                for words in instruction.to_words() {
+                    self.module.logical_layout.global_variables.push(words);
+                }
+                id
+            }
+        }
+    }
+
+    fn get_function_type(
+        &mut self,
+        ty: Option<crate::Handle<crate::Type>>,
+        arena: &crate::Arena<crate::Type>,
+    ) -> Word {
+        match ty {
+            Some(handle) => self.get_type_id(arena, &handle),
+            None => match self.void_type {
+                Some(id) => id,
+                None => {
+                    let id = self.generate_id();
+
+                    let mut instruction = Instruction::new(Op::TypeVoid);
+                    instruction.set_result(id);
+
+                    self.void_type = Some(id);
+                    for word in instruction.to_words().iter() {
+                        self.module.logical_layout.type_declarations.push(*word);
+                    }
+                    id
+                }
+            },
+        }
+    }
+
+    fn find_scalar_handle(
+        &self,
+        arena: &crate::Arena<crate::Type>,
+        kind: &crate::ScalarKind,
+        width: &u8,
+    ) -> crate::Handle<crate::Type> {
+        let mut scalar_handle = None;
+        for (handle, ty) in arena.iter() {
+            match ty.inner {
+                crate::TypeInner::Scalar {
+                    kind: _kind,
```

Comment:
would be better to rename function arguments to avoid collision with these
underscores should be left for things that are not used in Rust